Clarify the philosophy and intended uses for engines (#69)
ember-engines/ember-engines.com: markdown/guide/what-are-engines.md

## What are Engines?

> Engines allow multiple logical applications to be composed together into a single application from the user's perspective. Engines are isolated and composable and share many features with normal Ember applications. A notable exception is that an [Engine](https://api.emberjs.com/ember/release/classes/Engine) requires a host application to boot it and provide a `Router` instance.

Reviewer: ```suggestion
Engines allow multiple logical applications to be composed together into a single application from the user's perspective. The key term here is "logical application". That is, an Engine represents a set of functionality and user experiences that could logically be considered an application. However, they differ from normal Ember applications in that they are intended to be composed within a host application in order to create a single, cohesive user experience.

On a technical level, engines and applications are remarkably similar: they each have their own container, registry, context, and namespace. However, unlike an application, an engine does not boot itself nor does it control its own router. Engines rely on their hosts to fulfill these responsibilities.
```

> Engines allow multiple logical applications to be composed together into a single application from the user's perspective, that provide functionality to their host applications. Engines are isolated, `composable applications`, they have almost all the same features as normal Ember applications, except an [Engine](https://api.emberjs.com/ember/release/classes/Engine) requires a host application to boot it and provide a Router instance.

Reviewer: ```suggestion
Engines allow multiple logical applications to be composed together into a single application from the user's perspective. Engines are isolated and composable and share many features with normal Ember applications. A notable exception is that an [Engine](https://api.emberjs.com/ember/release/classes/Engine) requires a host application to boot it and provide a `Router` instance.
```

## What are Engines?

Engines can be considered miniature applications that provide functionality to their host applications. Engines are isolated, `composable applications`, they have almost all the same features as normal Ember applications, except an [Engine](https://api.emberjs.com/ember/release/classes/Engine) requires a host application to boot it and provide a Router instance.

> "Engines allow multiple logical applications to be composed together into a single application from the user's perspective." - [Engines Guides](http://ember-engines.com/)

Reviewer: I don't think we need to quote the engines guides from within the guides themselves. I would drop the quotes and replace move this paragraph to be the first sentence of the above paragraph (and drop the sentence with "miniature applications" in it.

Author: fixed

The often used example of this is a "blogging" application. While a blog itself could be a full-fledged application, you may also wish to use it as a specific subsection of your overall application. If you threw a "news feed" engine into that mix, you might get an application structure that looks something like this:

![Diagram of Application with two Engines](../images/02-app-diagram.png)

> Even though the application has sets of functionality represented by Engines, there is room for some additional functionality that the host application itself provides.

Reviewer: ```suggestion
This host application relies upon Engines to control different "zones" of functionality. However, from the user's perspective, the host application and its engines appear to be a single application.
```

## Why use Engines?

> Large companies are increasingly adopting Ember.js to power their entire product lines. Often this means separate teams (sometimes distributed around the world) working on the same app. Typically, responsibility is shared by dividing the application into one or more "sections". How this division is actually implemented varies from team to team. 

Reviewer: We need to avoid assuming "companies" and think in terms of "organizations".

```suggestion
Large organizations often use Ember.js to power sophisticated web applications. These apps may require collaboration among several teams, sometimes distributed around the world. Typically, responsibility is shared by dividing the application into one or more "sections". How this division is actually implemented varies from team to team. 
```

> These companies also have large monolithic applications, which pose the following challenges:

Reviewer: ```suggestion
Maintaining large monolithic applications poses the following challenges:
```

* `Side effects` - if you change something, it may be unclear how it could affect the rest of platform.
* `Coordination` - when you develop a new feature or make big changes, many teams may need to be in sync to approve it.
* `Complexity` - with a huge dependency tree and many layers of abstraction, developers cannot iterate quickly, and features suffer as a result.
* `Killing Innovation` - a/b testing a cutting-edge feature is hard to do without disrupting the rest of the app and the teams working on it.
* `Slow Onboarding` - new people coming into the team are overwhelmed.

> Engines provide an alternative to these approaches that allows for distributed development, testing, and packaging. The goal of Ember Engines was to allow large Ember applications to be split into consumable Addons allowing development teams to build logically-grouped pieces of an application.

Reviewer: ```suggestion
Engines provide an antidote to these problems by allowing for distributed development, testing, and packaging of logically-grouped pieces of an application.
```

> Engines are good for companies that have a large number of teams, where each team has their own area that is clearly separated from the others. If your app is developed by a small team, engines are probably not very helpful for you. The isolation is good when it matches organizational boundaries, but adds unnecessary complexity when there is not a good match.

Reviewer: I find this paragraph overly prescriptive and would probably just remove it. Who are we to say that engines couldn't be useful for two different developers working independently on a single application?

Author: @dgeb this is the same info on official ember guides approved by core team - https://guides.emberjs.com/release/applications/ember-engines/

And we're just saying it's probably not good, we're not giving a certainty, but almost always this seems to be overengineering - "engines are probably not very helpful for you."

Reviewer: @villander ok, I agree they should be consistent, but I really think this should be toned down in both places.  I feel that this recommendation is heavy handed. Saying "large number of teams" is overboard. It's fine to say multiple teams. And sometimes those are "teams" of one. The point being that Engines are an organizational solution.

Reviewer: And we need to say "organizations" and not "companies".

Author: makes sense, I'll change it here, and on official guides on the future.

Author: fixed @dgeb

> Engines are used by a number of companies, including [LinkedIn](linkedin.com), [Chase](https://www.chase.com/) and [Square](https://squareup.com) in production sites with million of users.

Reviewer: ```suggestion
Engines are used by a number of organizations, including [LinkedIn](linkedin.com), [Chase](https://www.chase.com/), and [Square](https://squareup.com), to power sites with million of users.
```
